packages/next/server/base-http.ts

```diff
@@ -0,0 +1,273 @@
+import type { ServerResponse, IncomingMessage } from 'http'
+import type { Writable, Readable } from 'stream'
+import { parseNextUrl } from '../dist/shared/lib/router/utils/parse-next-url'
+import { PERMANENT_REDIRECT_STATUS } from '../shared/lib/constants'
+import { ParsedNextUrl } from '../shared/lib/router/utils/parse-next-url'
+import { parseBody } from './api-utils'
+import { I18NConfig } from './config-shared'
+
+export interface BaseNextRequestConfig {
+  basePath: string | undefined
+  i18n?: I18NConfig
+  trailingSlash?: boolean | undefined
+}
+
+interface RequestMeta {
+  // interface from `server/request-meta.ts`
+}
+
+export abstract class BaseNextRequest<Body> {
+  constructor(
+    public url: string,
+    public method: string,
+    public body: Body,
+    public config: BaseNextRequestConfig
+  ) {}
+
+  abstract parseBody(limit: string | number): Promise<any>
+
+  abstract getHeader(name: string): string | string[] | undefined
+
+  abstract getAllHeaders(): Record<string, string | string[]>
+
```

Comment:
Might also need to add abstract methods for getting a cookie/all cookies. I know that in the web side, we'll have to do some extra implementation for `Set-Cookie` headers, so we might also have to inject some cookies into the request, but not 100% sure if needed yet

Reply:
I think that's already handled by https://github.com/vercel/next.js/blob/0f7f3323c180ebe7046187f3e338a8d818405ea7/packages/next/server/base-server.ts#L370 so might not needed in the base class right now.